src/Adapter/Product/ProductUpdater.php

```diff
@@ -91,15 +113,94 @@ public function refreshProductCustomizabilityProperties(Product $product): void
         );
     }
 
+    /**
+     * @param Product $product
+     *
+     * @throws CoreException
+     */
+    public function resetDefaultSupplier(Product $product): void
+    {
+        $this->update($product, [
+            'supplier_reference' => '',
+            'wholesale_price' => 0,
+            'id_supplier' => 0,
+        ], CannotUpdateProductException::FAILED_UPDATE_DEFAULT_SUPPLIER);
+    }
+
+    /**
+     * @param Product $product
+     * @param int $defaultSupplierId
+     */
+    public function updateDefaultSupplier(Product $product, int $defaultSupplierId): void
+    {
+        if ($defaultSupplierId && !$product->hasCombinations()) {
+            $this->resetDefaultSupplierIfNotExists($product, new SupplierId($defaultSupplierId));
+            $this->update($product, [
+                'supplier_reference' => ProductSupplier::getProductSupplierReference($product->id, 0, $defaultSupplierId),
+                'wholesale_price' => ProductSupplier::getProductSupplierPrice($product->id, 0, $defaultSupplierId),
+                'id_supplier' => $defaultSupplierId,
+            ], CannotUpdateProductException::FAILED_UPDATE_DEFAULT_SUPPLIER);
+
+            return;
+        }
+
+        $this->resetDefaultSupplier($product);
+    }
+
+    /**
+     * @param Product $product
+     * @param SupplierId $supplierId
+     *
+     * @throws CoreException
+     * @throws ProductSupplierNotFoundException
+     * @throws SupplierNotFoundException
+     */
+    private function resetDefaultSupplierIfNotExists(Product $product, SupplierId $supplierId): void
+    {
+        try {
+            $this->supplierProvider->assertSupplierExists($supplierId);
+            $this->productSupplierProvider->assertProductSupplierExists(new ProductSupplierId(
+                (int) ProductSupplier::getIdByProductAndSupplier(
+                    $product->id,
+                    0,
+                    $supplierId->getValue())
+            ));
+        } catch (SupplierNotFoundException | ProductSupplierNotFoundException $e) {
+            $this->resetDefaultSupplier($product);
```

Comment:
I don't think this is a good idea, because it will erase the previous default supplier So if a product has a valid default supplier, and you intend to set one that doesn't exists It will not only fail the update but remove the previous default supplier
It should only checks if the supplier exists, if it doesn't it simply throws an error and nothing is changed

One thing we could do however is check if the supplier is correctly associated to the Product, and if not associate it when one wants to set it as default (this would avoid the problem you mentioned earlier: what happens if you try to use a default supplier that was not previously associated)